test/exoscale/compute/api/http_test.clj

```diff
@@ -1,29 +1,85 @@
-           (http/find-payload {:jobid "dac8bd04-31ff-11ea-ba11-6bef69b2ab5f"}))))
+           (http/find-payload {:jobid "dac8bd04-31ff-11ea-ba11-6bef69b2ab5f"} :someMutation))))
   (testing "vector"
     (is (= [{:id "bar"}]
-           (http/find-payload {:network [{:id "bar"}]
-                               :count 1})))))
+           (http/find-payload {:network [{:id "bar"}] :count 1} :someMutation)))))
+
+(defn json-request-spy [items]
+  (spy/spy (fn [config opcode {:keys [page pagesize]}]
+             (d/success-deferred (with-meta (->> items
+                                                 (drop (* (dec page) pagesize))
+                                                 (take pagesize)
+                                                 (map (fn [n]
+                                                        {:item n})))
+                                   {:count (count items)})))))
+
+(deftest pagination-no-params-test
+  (let [spy (json-request-spy (range 0 5))]
+    (with-redefs [http/json-request!! spy]
+      (let [res @(http/list-request!! {} "listZones" {})]
+        (is (= [{:item 0} {:item 1} {:item 2} {:item 3} {:item 4}] res))
+        (is (= 5 (:count (meta res))))
+        (assert/called-once-with? spy {} "listZones" {:page 1 :pagesize 500})))))
+
+(deftest pagination-first-page-test
+  (let [spy (json-request-spy (range 0 5))]
+    (with-redefs [http/json-request!! spy]
+      (let [res @(http/list-request!! {} "listZones" {:page 1 :pagesize 2})]
+        (is (= [{:item 0} {:item 1}] res))
+        (is (= 5 (:count (meta res))))
+        (assert/called-once-with? spy {} "listZones" {:page 1 :pagesize 2})))))
+
+(deftest pagination-second-page-test
+  (let [spy (json-request-spy (range 0 5))]
+    (with-redefs [http/json-request!! spy]
+      (let [res @(http/list-request!! {} "listZones" {:page 2 :pagesize 2})]
+        (is (= [{:item 2} {:item 3}] res))
+        (is (= 5 (:count (meta res))))
+        (assert/called-once-with? spy {} "listZones" {:page 2 :pagesize 2})))))
+
+(deftest pagination-last-page-test
+  (let [spy (json-request-spy (range 0 5))]
+    (with-redefs [http/json-request!! spy]
+      (let [res @(http/list-request!! {} "listZones" {:page 3 :pagesize 2})]
+        (is (= [{:item 4}] res))
+        (is (= 5 (:count (meta res))))
+        (assert/called-once-with? spy {} "listZones" {:page 3 :pagesize 2})))))
+
+(deftest pagination-max-500-test
+  (let [spy (json-request-spy (range 0 600))]
+    (with-redefs [http/json-request!! spy]
+      (let [res @(http/list-request!! {} "listZones" {})]
+        (is (= 500 (count res)))
+        (is (= 600 (:count (meta res))))
+        (assert/called-once-with? spy {} "listZones" {:page 1 :pagesize 500}))
+
+      (spy/reset-spy! spy)
+      (let [res @(http/list-request!! {} "listZones" {:page 2})]
+        (is (= 100 (count res)))
+        (is (= 600 (:count (meta res))))
+        (assert/called-once-with? spy {} "listZones" {:page 2 :pagesize 500})))))
```

Comment:
I wasn't sure what the default behaviour should do here, whether we always stop at 500 and require the user to make an additional request to get the next page - or whether we do this for them in the `d/loop`? if so this needs to be changed so we loop and we get all 600 results back - but make 2 calls to the spy.

Reply:
We do it in the loop, list-request!! pages through all results

Reply:
Fixed in https://github.com/exoscale/clojure-exoscale/pull/15/commits/9eff65e9e0bed147bcec48230ef4f2465b3d61a9